src/big/big_decimal.cr

```diff
@@ -0,0 +1,241 @@
+  end
+
+  def /(other : BigDecimal) : BigDecimal
+    check_division_by_zero other
+
+    scale = @scale - other.scale
+    n, d = @value, other.@value
+
+    quotient, remainder = n.tdiv(d), n.remainder(d)
+    return BigDecimal.new(quotient, scale) if remainder == ZERO
+
+    remainder = remainder * TEN
+
+    i = 0
+    while remainder != ZERO && i < @@max_div_iterations
+      q, r = remainder.tdiv(d), remainder.remainder(d)
+      quotient = quotient * TEN + q
+      remainder = r * TEN
+      i += 1
+    end
+
+    if (@value < 0 && other.value > 0) || (other.value < 0 && @value > 0)
+      BigDecimal.new(-quotient, scale + i)
+    else
+      BigDecimal.new(quotient, scale + i)
+    end
+  end
+
+  def <=>(other : BigDecimal) : Int32
+    if @scale > other.scale
+      @value <=> other.scale_to(self).value
+    elsif @scale < other.scale
+      scale_to(other).value <=> other.value
+    else
+      @value <=> other.value
+    end
+  end
+
+  def <=>(other : Int)
+    @value <=> other
+  end
+
+  def ==(other : BigDecimal) : Bool
+    if @scale > other.scale
+      scaled = other.value * power_ten_to(@scale - other.scale)
+      @value == scaled
+    elsif @scale < other.scale
+      scaled = @value * power_ten_to(other.scale - @scale)
+      scaled == other.value
+    else
+      @value == other.value
+    end
+  end
+
+  # Scales a BigDecimal to another BigDecimal, so they can be
+  # computed easier.
+  def scale_to(new_scale : BigDecimal) : BigDecimal
+    new_scale = new_scale.scale
+
+    if @value == 0
+      BigDecimal.new(0.to_big_i, new_scale)
+    elsif @scale > new_scale
+      scale_diff = @scale - new_scale.to_big_i
+      BigDecimal.new(@value / power_ten_to(scale_diff), new_scale)
+    elsif @scale < new_scale
+      scale_diff = new_scale - @scale.to_big_i
+      BigDecimal.new(@value * power_ten_to(scale_diff), new_scale)
+    else
+      self
+    end
+  end
+
+  def to_s(io : IO)
+    s = @value.to_s
+
+    if @scale == 0
+      io << s
+      return
+    end
+
+    if @scale == s.size
+      io << "0." << s
+    elsif @scale > s.size
+      io << "0."
+      (@scale - s.size).times do
+        io << "0"
+      end
+      io << s
+    else
+      offset = s.size - @scale
+      io << s[0...offset] << "." << s[offset..-1]
+    end
+  end
+
+  private def check_division_by_zero(bd : BigDecimal)
+    raise DivisionByZero.new if bd.value == 0
+  end
+
+  private def power_ten_to(x : Int) : Int
+    TEN ** x
+  end
+end
+
+struct Int
+  include Comparable(BigDecimal)
+
+  # Convert Int to BigDecimal
+  def to_big_d
+    BigDecimal.new(self)
+  end
+
+  def <=>(other : BigDecimal)
+    self <=> other.value
+  end
+end
+
+class String
+  include Comparable(BigDecimal)
+
+  # Convert String to BigDecimal
+  def to_big_d
+    BigDecimal.new(self)
+  end
+end
+
+struct Float
+  # from Float is not supported due to precision loss risks. This call fails at compile time.
```

Comment:
@vegai you should use `` ` `` around `Float` in your comment: ``Casting from `Float` ...``